agent/pkg/proxy/socks5_proxy.go

```diff
@@ -0,0 +1,145 @@
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"github.com/kubeedge/edgemesh/agent/pkg/proxy/protocol"
+	"github.com/kubeedge/edgemesh/agent/pkg/tunnel"
+	"github.com/kubeedge/edgemesh/common/constants"
+)
+
+type Socks5Proxy struct {
+	TCPProxy   *protocol.TCPProxy
+	kubeClient kubernetes.Interface
+	NodeName   string
+}
+
+const (
+	labelKubeedge string = "kubeedge=edgemesh-agent"
+	agentPodName  string = "edgemesh-agent"
+)
+
+func (s *Socks5Proxy) Start() {
+	go func() {
+		for {
+			conn, err := s.TCPProxy.Listener.Accept()
+			if err != nil {
+				klog.Warningf("get socks5 tcp conn error: %v", err)
+				continue
+			}
+			go s.HandleSocksProxy(conn)
+		}
+	}()
+}
+
+func NewSocks5Proxy(ip net.IP, port int, NodeName string, kubeClient kubernetes.Interface) (socks5Proxy *Socks5Proxy, err error) {
+	socks := &Socks5Proxy{
+		kubeClient: kubeClient,
+		TCPProxy:   &protocol.TCPProxy{Name: protocol.TCP},
+		NodeName:   NodeName,
+	}
+
+	if err := socks.TCPProxy.SetListener(ip, port); err != nil {
+		return socks, fmt.Errorf("set socks5 proxy err: %v, host: %s, port: %d", err, ip, port)
+	}
+
+	return socks, nil
+}
+
+func pipe(dst io.WriteCloser, src io.ReadCloser, closeOnce *sync.Once) {
+	_, err := io.Copy(dst, src)
+	if err != nil && err != io.EOF && !strings.Contains(err.Error(), constants.ConnectionClosed) && !strings.Contains(err.Error(), constants.StreamReset) {
+		klog.Errorf("io copy between proxy and client error: %v", err)
+	}
+	closeOnce.Do(func() {
+		dst.Close()
+		src.Close()
+	})
+}
+
+func (s *Socks5Proxy) HandleSocksProxy(conn net.Conn) {
+	if conn == nil {
+		return
+	}
+	defer conn.Close()
+
+	var b [1024]byte
+	n, err := conn.Read(b[:])
+	if err != nil {
+		klog.Errorf("Unable to get data from client connection, err: %v", err)
+		return
+	}
+
+	if b[0] == 0x05 {
```

Comment:
Should we consider defining variables instead of hard coding here?

Reply:
Thank you for your suggestion. I have optimized the code.